src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs

```diff
@@ -101,6 +140,14 @@ internal sealed override bool OnTryRead(ref Utf8JsonReader reader, Type typeToCo
 
                 obj = (T)CreateObject(ref state.Current);
 
+                if (state.Current.MetadataPropertyNames.HasFlag(MetadataPropertyName.Id))
+                {
+                    Debug.Assert(state.ReferenceId != null);
+                    Debug.Assert(options.ReferenceHandlingStrategy == ReferenceHandlingStrategy.Preserve);
+                    state.ReferenceResolver.AddReference(state.ReferenceId, obj);
+                    state.ReferenceId = null;
```

Comment:
should you be also removing `Id` flag here? I think you should create a routines for these so that these lines of code are consistent everywhere

Reply:
Removing the flag is not necessary (since it's stored in the stack frame rather than the parent state object). I considered extracting to a method, however the logic is slightly different for objects and collections so it didn't provide a lot of benefit.

Reply:
Well, fields defined in `ReadState` are unique in the deserialization context, whereas fields defined in `ReadStackFrame` are replicated for every value that is pushed in the stack -- there's no need to reset the latter kind.